### PR TITLE
FAO Publications: Update DOI method

### PR DESCRIPTION
**Pull request by swifterslb**
Reupload for https://github.com/zotero/translators/pull/2191 

DOI has become standard for new FAO publications.
Method for obtaining the DOI is optimized. Also assign DOI URL as newItem.url when DOI exists.

Thanks!

### CODE REVIEW

**Comment by dstillman** [comment]
You don't need to, and shouldn't, create a new PR. You need to rebase against master and resolve conflicts with that branch.

**Comment by dstillman** [comment]
No, that's wrong. You need to rebase against the master branch of upstream/translators, such that only your own commits for this PR are shown. There are lots of sources online explaining how to do that.

**Comment by swifterslb** [comment]
I'm sorry I just got some time to work on this. I think I have successfully rebased `swifterslb-patch-2` branch to `master` and updated the file `FAO Publications.js`. Travis CI is showing errors for some other files. Could you please have a look? Thanks and sorry for all the trouble I may have caused.

**Comment by dstillman** [comment]
No, you definitely have not rebased it — there are 19 other files included in this commit. You need to rebase so that you only see changes to `FAO Publications.js` locally before force-pushing.

**Comment by swifterslb** [comment]
> No, you definitely have not rebased it — there are 19 other files included in this PR. You need to rebase against the master branch of the main repo (possibly `upstream master`, depending on how you set up your remotes) so that you only see changes to `FAO Publications.js` locally before force-pushing.
> 
> We can't really help with this, but there are lots of resources online explaining how to do an interactive rebase.

I feel really stupid... I couldn't get my head around Git BASH for interactive rebase, and then I found GitKraken which seems to be easier, but I still couldn't _interactively_ rebase `swifterslb-patch-2` branch to `zotero/master`, maybe because it was too messed up... so I hard reset the branch to `zotero/master` (which I didn't know would cause the the pull request to close), reupdated the branch, and reopened the pull request. Thanks for the patience.

**Comment by dstillman** [comment]
This is still wrong. We can't review this until it's a clean patch, rebased on a fully fetched upstream/master.

**Comment by AbeJellinek** [comment]
You're not stupid. Git is just hard. I just rebased and pushed:

```bash
$ git fetch upstream          # fetch the zotero/translators repository
$ git rebase upstream/master  # rebase onto its master branch
$ git push --force            # push, confirming that you want to overwrite the remote history
```

To pull the rebased branch, you'll want to run `git pull --force`.

File: FAO Publications.js [hunk 1]
```diff
@@ -153,7 +158,7 @@ function scrape(doc, url) {
 		if (!subTitle) {
 			newItem.title = mainTitle;
 		}
-		else if ((newItem.language == 'zh') || (newItem.language == 'ja') || (newItem.language == 'ko')) {
+		else if ((newItem.language == 'zh') || (newItem.language == 'ja')) {
```
Comment (AbeJellinek) [inline FAO Publications.js:161]: Why exclude Korean? Doesn't seem relevant to DOI changes.

File: FAO Publications.js [hunk 2]
```diff
@@ -98,10 +98,13 @@ function scrape(doc, url) {
...
-			// DOI: Some docs contain DOI as the last paragraph in abs field
+			// DOI: Some docs contain DOI as a separate paragraph in abs field
+			// use DOI URL as url when DOI exists
 			var DOILead = 'https://doi.org/';
-			if (abs.textContent.includes(DOILead)) {
-				newItem.DOI = abs.textContent.slice(abs.textContent.indexOf(DOILead) + DOILead.length);
+			if (abs.innerText.includes(DOILead)) {
+				var DOIMatch = abs.innerText.match(/https:\/\/doi\.org\/(.+)/i);
+				newItem.DOI = DOIMatch[1];
+				newItem.url = DOIMatch[0];
```
Comment (AbeJellinek) [inline FAO Publications.js:107]: I don't think we should use the DOI as the item's URL, for three reasons:

1. `http://www.fao.org/documents/**` URLs (yes, shockingly, no HTTPS whatsoever on a UN site) are the preferred permalinks according to the site.
2. If FAO.org ever goes down, its pages will still be saved in the Wayback Machine. If DOI.org ever goes down, *its* pages, which are just redirects, mostly won't be.
3. We already store the DOI in the DOI field. Also storing it in the URL field won't provide any additional information.

File: FAO Publications.js [hunk 3]
```diff
@@ -111,8 +114,10 @@ function scrape(doc, url) {
 			title: 'Full Text PDF',
 			mimeType: 'application/pdf'
 		});
-		// url
-		newItem.url = url;
+		// url when DOI doesn't exist
+		if (!abs.innerText.includes(DOILead)) {
```
Comment (AbeJellinek) [inline FAO Publications.js:118]: As above, revert this change.

File: FAO Publications.js [hunk 4]
```diff
@@ -359,7 +416,7 @@ var testCases = [
 				"publisher": "FAO",
 				"series": "FAO Fisheries and Aquaculture Circular",
 				"seriesNumber": "No. 1207",
-				"url": "http://www.fao.org/documents/card/en/c/ca8751en/",
```
Comment (AbeJellinek) [inline FAO Publications.js:362]: ...and update tests to match.

**Review by AbeJellinek — CHANGES_REQUESTED** [review]

**Comment by AbeJellinek** [comment]
OK, left a few comments. I promise we can get this merged quickly!

File: FAO Publications.js (hunk 1 above)
Comment (swifterslb) [reply FAO Publications.js:161]: No this is not related to DOI changes. I confirmed that Korean uses half-width punctuations (":"), not full-width ones like Chinese and Japanese ("："), so I removed it.

File: FAO Publications.js (hunk 2 above)
Comment (swifterslb) [reply FAO Publications.js:107]: I use the DOI link (when possible) for the URL because from 2021 FAO recommends using the DOI link in citation. All FAO publications in 2021, in their "recommended citation" or "required citation" section, use DOI link, e.g.:

* http://www.fao.org/publications/card/en/c/CB3913EN
* http://www.fao.org/publications/card/en/c/CB1841EN
* http://www.fao.org/publications/card/en/c/CB3196EN

The citation text is at the "Cite this content as" section on the web page and the copyright page of the PDF file (click the PDF icon on the right of " Download", usually at the 4th page). For all publications before 2021 (when DOI was not generated for publications) the URL scheme remains unchanged.

By assigning DOI as the item's URL, Zotero can automatically generate the correct citation text when using the "Food and Agriculture Organization of the United Nations" style. This saves a tremendous amount of drudgeries for FAO writers and editors, which was indeed the primary goal I developed this Zotero translator and the corresponding style file.

I fully understand this practice is not in full conformity with the intended usage of Zotero, but I'd really appreciate it if you could let us use it this way.

**Comment by swifterslb** [comment]
> OK, left a few comments. I promise we can get this merged quickly!

Thanks so much for taking this up again and helping me out! Please consider our situation and make a decision on whether I could assign DOI link as URL, and I'll adjust accordingly.

File: FAO Publications.js (hunk 2 above)
Comment (AbeJellinek) [reply FAO Publications.js:107]: [The style](https://github.com/citation-style-language/styles/blob/2cd6364de42525f99d00b99d790603643ec67a52/food-and-agriculture-organization-of-the-united-nations.csl#L212) as it exists in the CSL styles repo actually handles journal articles correctly already - if an article has both a DOI and a URL, it'll just show the DOI (as a doi.org link). It does not do that for other item types like books. Fixing that should be easy, and it'll let us avoid coupling a translator with a citation style.

**Comment by AbeJellinek** [comment]
No problem! See my comment in the discussion above.

**Comment by swifterslb** [comment]
Thanks so much for the great suggestion. I'll update the CSL file later.

For the failing check: Due to limitations of this site, items must be assigned as "book" type first. If it's a "conferencePaper" type, it needs to be assigned later. Test case 6 is correctly assigned the "conferencePaper" type. Could you please solve this?

File: FAO Publications.js [hunk 5]
```diff
@@ -169,10 +170,10 @@ function scrape(doc, url) {
 		var metaText = ZU.xpath(doc, '//*[@id="mainN0"]')[0].innerText.split('\n'); // scrape text of meta area and split into an array based on line breaks.
 		// get what variables are listed in the page, save to object existingMeta
 		var textVariable = { // declarations for metadata names as appeared in document pages in different languages
-			date: ['سنة النشر', '出版年代', 'Year of publication', 'Année de publication', 'Год издания', 'Fecha de publicación'],
+			date: ['سنة النشر', '出版年份', 'Year of publication', 'Année de publication', 'Год издания', 'Fecha de publicación'],
```
Comment (swifterslb) [inline FAO Publications.js:173]: Some changes of metadata terms in Chinese.

**Comment by AbeJellinek** [comment]
It looks like you opened the PR in your own fork, not the official CSL styles repo, but otherwise looks good to me.

**Comment by swifterslb** [comment]
> It looks like you opened the PR in your own fork, not the official CSL styles repo, but otherwise looks good to me.

Thanks. I've made the PR to official CSL styles repo.

**Comment by AbeJellinek** [comment]
No problem! Regarding determining whether it's a `conferencePaper` or `book` in `detectWeb`, it looks like looping through elements with class `fdr_label` and checking whether one includes the text `Meeting Name:` should be sufficient. Could just do a for loop, or something like

```js
let isConferencePaper = [...doc.querySelectorAll('.fdr_label')]
	.some(label => label.innerText.includes('Meeting Name:'));
```

File: FAO Publications.js [hunk 6]
```diff
@@ -535,8 +607,8 @@ var testCases = [
 				"numPages": "251",
 				"place": "Rome, Italy",
 				"publisher": "FAO",
-				"series": "Non-wood forest products working paper",
-				"seriesNumber": "21",
+				"series": "Produits Forestiers Non-Ligneux",
+				"seriesNumber": "No. 21",
```
Comment (AbeJellinek) [inline FAO Publications.js:611]: Ack, this is a regression (probably because you're scraping the French site and the tests were originally created on the English one, so nobody caught it). Can we strip the "No." prefix? Maybe just match to find the numeric part so that it works across languages, like

```js
newItem.seriesNumber = (newItem.seriesNumber.match(/\d+/) || [])[0];
```

That will pull out the number if there is one, or `undefined` if there isn't.

**Review by AbeJellinek — CHANGES_REQUESTED** [review]
OK, one last comment, sorry. Otherwise I think this is ready.

**Comment by AbeJellinek** [comment]
Thanks so much for bearing with all the suggestions. Merged.

**Comment by swifterslb** [comment]
> Thanks so much for bearing with all the suggestions. Merged.

Thank you so much for helping me out! I've learned a lot.
